### ISSUE TITLE
publish devpi server as container image

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [devpi/devpi#439]
in terms of setup a containerized version of the image thats using a data volume is way more easy to setup than installing a local service thats managing local folders

in order to properly support that however there are a few details that need handling

- [ ] data upgrade paths (my proposal would be to fail on incompatible data formats and support import/export in a distinct way via passing arguments)
- [ ] handling of replication setups and their linking on different nodes of a setup

**Comment by EvaSDK** [comment]
Similar to #428

**Comment by yunfan** [comment]
still open?

**Comment by fschulze** [comment]
No one stepped up to do it yet.

**Comment by RonnyPfannschmidt** [comment]
even at work we currently only have basic setups that can't yet handle migration/update paths well

**Comment by yunfan** [comment]
i had turn to squid in a docker solution to solve the same problem now, i think most user's purpose is just to setup an mirror of the upstream

**Comment by JonasAlfredsson** [comment]
I have created an [image](https://github.com/JonasAlfredsson/docker-devpi#upgrading) that does some of the import/export work for you, but I have not yet found a simple method of fully automating this. 

I can't find it in the documentation, so I am unsure if a newer version of the server can "export" an older version before "importing" again. Most likely the old version needs to do the exporting before the new can do the importing, and then it is much more difficult to make the upgrade process smooth as we will need external input from an admin to tell the image when one of these processes should be carried out.

**Comment by fschulze** [comment]
You are right, the export has to be done by the old version. But it wasn't needed the past years. I expect the next major version to need an export/import cycle though, but it is still a few months away. I'm still trying to come up with a way to avoid needing export/import for that version, but if that takes too long or is too complex it will be needed.

**Comment by fschulze** [comment]
The last major releases didn't require it. It is only necessary if the database schema changes.

**Comment by fschulze** [comment]
And to say it the other way around, if the database schema changes it will definitely be in a major release, never a minor or patch release.

**Comment by JonasAlfredsson** [comment]
Excellent information, I updated my README :)

**Comment by fschulze** [comment]
These https://github.com/JonasAlfredsson/docker-devpi from https://github.com/devpi/devpi/issues/661#issuecomment-1064083271 seem to be fairly up to date and maintained.

**Comment by RonnyPfannschmidt** [comment]
looks lovely, unfortunately my setup at work was decommissioned a while back

**Comment by JonasAlfredsson** [comment]
I have added dependabot to watch for new devpi releases, so as long as a release is properly published to pypi dependabot  should pick it up and automatically update the requirements.txt files used by my Docker images. So they should (hopefully) always be up to date and pushed to DockerHub with the same version tags.

**Comment by AmirHmZz** [comment]
@fschulze @JonasAlfredsson Any motivation to make a PR to do all CI/CD stuff here in the official repo?

**Comment by JonasAlfredsson** [comment]
Personally I would want to keep the Docker stuff separate from this repo to make the organisation much simpler, so I have no intention of merging that code in here. 

However, if @fschulze want to just merge it in here he is more than welcome :) 
(Or just take over the other repo to make it more official). 

But since he is not using Docker I don't think he is that interested ^^
So I will keep my simple example up to date as long as things are published on pypi. :)

**Comment by fschulze** [comment]
As I don't use Docker myself, I'd like to keep it separate. I labeled this issues with ``Documentation`` now, to add a link to the repo to the docs if @JonasAlfredsson is ok with that.

**Comment by JonasAlfredsson** [comment]
Just edited my comment, but yeah, as long as updates end up on pypi Dependabpt will pick them up and I will merge them to keep the image up to date. 

Very little work for me at the moment, so I will keep maintaining it. You can link to it if you want :)

Ping me if you are planning some breaking changes so I have some time to prepare :P 

Also, the docs regarding how to handle MAJOR upgrades probably need to be streamlined a bit, but perhaps that is a task for when it actually happens ^^
